### ISSUE TITLE
t.p.components.implements() has confusing behavior

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11288]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#457|
|Type|defect|
|Created|2004-01-02 22:48:04Z|

<details><summary>Searchable metadata</summary>

```
trac-id__457 457
type__defect defect
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1073083684000000 1073083684000000
changetime__1470261724473708 1470261724473708
version__ 
owner__ 
cc__glyph cc__radix cc__exarkun cc__itamarst cc__warner cc__dash
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
There is no reliable way to indicate interfaces a class or type object implements:

class I1(Interface): pass
class I2(Interface): pass

class A(type):
  __implements__ = I1

class B(object):
  __metaclass__ = A
  __implements__ = I2

t.p.components.implements(B, I1) -> False

This is related to the also incorrect behavior:

t.p.components.implements(B, I2) -> True

which we also happen to have unit tests for.


Zope3 uses __class_implements__ to resolve this.  I don't particularly like
that, but it's better than what we have now.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Zope3 no longer has __implements__, so dunno how __class_implements__ is doing now.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph commented|
|-|-|

```
#!html
<pre>
I assume this will be fixed by the new component model upgrade?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Yes indeed.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Function is now deprecated.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|GitHub commented|
|-|-|

In changeset c5da63efa363fb30f4fd60af0c36ebcce2b2cdd4
```
#!CommitTicketReference repository="" revision="c5da63efa363fb30f4fd60af0c36ebcce2b2cdd4"
Merge pull request #457 from twisted/8722-einarfd-py3-conch-endpoints

Author: einarfd
Reviewer: rodrigc
Fixes: #8722 

Port twisted.conch.endpoints to Python 3
```
